### PR TITLE
fix RecursionError with synchronous deferreds

### PR DESCRIPTION
**Pull request by glyph**
## Scope and purpose

https://twistedmatrix.com/trac/ticket/10125

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10125
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

**Review by wsanchez — APPROVED** [review]
LGTM

File: src/twisted/web/_flatten.py [hunk 1]
```diff
@@ -324,44 +330,67 @@ def cbx(originalAndToFlatten):
...
-            result.callback(None)
-        except BaseException:
-            result.errback()
-        else:
+    Await each value in the iterable, completing when they're all done.
 
-            def cby(original):
-                _writeFlattenedData(state, write, result)
-                return original
+    @param iterator: An iterator of L{Deferred}s
```
Comment (pwestlak-xilinx) [inline src/twisted/web/_flatten.py]: The parameter is called `iterable`, not `iterator`.

**Review by pwestlak-xilinx — APPROVED** [review]
If I understand it correctly, this looks fine.

File: src/twisted/web/_flatten.py [hunk 2]
```diff
@@ -157,6 +157,25 @@ def _getSlotValue(name, slotData, default=None):
...
+    """
+    d2 = Deferred()
+
+    def callback(result):
+        d2.callback(result)
+        return result
+
+    def errback(failure):
+        d2.errback(failure)
+        return failure
```
Comment (glyph) [inline src/twisted/web/_flatten.py:173]: Thanks codecov. This uncovered line actually illustrates a really significant coverage gap in the previous behavior. I might have to do a test-refactoring PR first to fix this.

**Comment by mthuurne** [comment]
There is a problem in the way `_fork()` handles cancellation:
```
>>> from twisted.internet.defer import Deferred
>>> from twisted.web._flatten import _fork
>>> _fork(Deferred()).cancel()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mth/src/twisted/src/twisted/internet/defer.py", line 728, in cancel
    canceller(self)
TypeError: cancel() takes 1 positional argument but 2 were given
```

**Pull request merged**